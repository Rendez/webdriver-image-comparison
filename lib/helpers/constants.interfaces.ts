export interface AndroidOffsets {
  [key: number]: {
    // The height of the status bar
    STATUS_BAR: number;
    // The height of the address bar
    ADDRESS_BAR: number;
    // The height of the tool bar
    TOOL_BAR: number;
  };
}

export interface IosOffsets {
  [key: number]: {
    // The height of the status bar
    STATUS_BAR: number;
    // The height of the status bar for an iPad Pro
    STATUS_BAR_PRO: number;
    // The height of the status bar for an iPhone X#
    STATUS_BAR_X: number;
    // The height of the address bar
    ADDRESS_BAR: number;
    // The home bar data
    HOME_BAR: {
      // The default home bar data
      DEFAULT: {
        // The height of the home bar
        height: number;
        // The width of the home bar
        width: number;
        // The x position of the home bar
        x: number;
        // The y position of the home bar
        y: number;
      };
      // The large home bar data
      LARGE: {
        // The height of the home bar
        height: number;
        // The width of the home bar
        width: number;
        // The x position of the home bar
        x: number;
        // The y position of the home bar
        y: number;
      };
    };
  };
}
