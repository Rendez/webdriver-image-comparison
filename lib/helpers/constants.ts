import { IosOffsets } from './constants.interfaces';

export const DEFAULT_FORMAT_STRING = '{tag}-{browserName}-{width}x{height}-dpr-{dpr}';
export const PLATFORMS = {
  ANDROID: 'android',
  IOS: 'ios',
};
export const FOLDERS = {
  ACTUAL: 'actual',
  DIFF: 'diff',
  TEMP_FULL_SCREEN: 'tempFullScreen',
  DEFAULT: {
    BASE: './wic/baseline/',
    SCREENSHOTS: '.tmp/',
  },
};
export const DEFAULT_SHADOW = {
  ADDRESS_BAR: 6,
  TOOL_BAR: 6,
};
export const DESKTOP = 'desktop';
export const CUSTOM_CSS_ID = 'pic-css';
export const DEFAULT_RESIZE_DIMENSIONS = {
  top: 0,
  right: 0,
  bottom: 0,
  left: 0,
};
export const FULL_PAGE_SCROLL_TIMEOUT = 1500;
export const DEFAULT_TABBABLE_OPTIONS = {
  circle: {
    backgroundColor: '#ff0000',
    borderColor: '#000',
    borderWidth: 1,
    fontColor: '#fff',
    fontFamily: 'Arial',
    fontSize: 10,
    size: 10,
    showNumber: true,
  },
  line: {
    color: '#000',
    width: 1,
  },
};

/**
 * Android and iOS offsets data
 */
const ANDROID_DEFAULT_OFFSETS = {
  STATUS_BAR: 24,
  ADDRESS_BAR: 56,
  TOOL_BAR: 48,
};
export const ANDROID_OFFSETS = {
  6: { ...ANDROID_DEFAULT_OFFSETS },
  7: { ...ANDROID_DEFAULT_OFFSETS },
  8: { ...ANDROID_DEFAULT_OFFSETS },
  9: { ...ANDROID_DEFAULT_OFFSETS },
  10: { ...ANDROID_DEFAULT_OFFSETS },
  11: { ...ANDROID_DEFAULT_OFFSETS },
  12: { ...ANDROID_DEFAULT_OFFSETS },
};

//=========================
// iPhone/iPad screen data
//=========================
// Height of screen in portrait mode, this is also what you get back
// when you:
//  - ask for the screen dimensions in WebdriverIO `getWindowSize`
//  - with `screen.height` in Safari devtools
//  - with http://whatismyscreenresolution.net/
//------------------------------------------------------------------
// iPhones
// 6/7/8/SE/2020:                                667 (DEFAULT)
// 6/7/8/Plus:                                   736 (PLUS)
// X|XS|11Pro|12Mini|13Mini (notch):             812 (SMALL)
// 12|12Pro|13|13Pro (notch):                    844 (MEDIUM)
// XSMax|XR|11|11ProMax (notch):                 896 (LARGE)
// 12ProMax|13ProMax (notch):                    926 (EXTRA_LARGE)
//------------------------------------------------------------------
// iPads
// 9.7|Air 1/2|Mini 1/2/3/4/2019|Pro 9.7:       1024 (DEFAULT)
// 10.2 1/2019/2020/2021                        1080 (IPAD_10_2)
// Pro 10.5/2017/2019:                          1112 (PRO_10_5)
// Mini 2021 (notch):                           1133 (SMALL)
// Air 2020/2022 (notch):                       1180 (MEDIUM)
// Pro 11 2018/2021 (notch):                    1194 (LARGE)
// Pro 12.9 2017 (no notch)/2018/2021 (notch):  1366 (EXTRA_LARGE)
//------------------------------------------------------------------
export const IOS_OFFSETS: IosOffsets = {
  IPHONE: {
    667: {
      LANDSCAPE: {
        STATUS_BAR: 0,
        ADDRESS_BAR: 50,
        HOME_BAR: { x: 0, y: 0, height: 0, width: 0 },
      },
      PORTRAIT: {
        STATUS_BAR: 20,
        ADDRESS_BAR: 50,
        HOME_BAR: { x: 0, y: 0, height: 0, width: 0 },
      },
    },
    736: {
      LANDSCAPE: {
        STATUS_BAR: 0,
        ADDRESS_BAR: 50,
        HOME_BAR: { x: 0, y: 0, height: 0, width: 0 },
      },
      PORTRAIT: {
        STATUS_BAR: 20,
        ADDRESS_BAR: 50,
        HOME_BAR: { x: 0, y: 0, height: 0, width: 0 },
      },
    },
    // With Notch
    812: {
      LANDSCAPE: {
        STATUS_BAR: 0,
        ADDRESS_BAR: 50,
        HOME_BAR: { x: 300, y: 359, height: 9, width: 212 },
      },
      PORTRAIT: {
        STATUS_BAR: 44,
        ADDRESS_BAR: 50,
        HOME_BAR: { x: 118, y: 796, height: 9, width: 137 },
      },
    },
    844: {
      LANDSCAPE: {
        STATUS_BAR: 0,
        ADDRESS_BAR: 50,
        HOME_BAR: { x: 312, y: 375, height: 9, width: 220 },
      },
      PORTRAIT: {
        STATUS_BAR: 47,
        ADDRESS_BAR: 50,
        HOME_BAR: { x: 124, y: 829, height: 9, width: 143 },
      },
    },
    896: {
      LANDSCAPE: {
        STATUS_BAR: 0,
        ADDRESS_BAR: 50,
        HOME_BAR: { x: 331, y: 399, height: 9, width: 234 },
      },
      PORTRAIT: {
        STATUS_BAR: 48,
        ADDRESS_BAR: 50,
        HOME_BAR: { x: 131, y: 881, height: 9, width: 152 },
      },
    },
    926: {
      LANDSCAPE: {
        STATUS_BAR: 0,
        ADDRESS_BAR: 50,
        HOME_BAR: { x: 342, y: 413, height: 9, width: 242 },
      },
      PORTRAIT: {
        STATUS_BAR: 47,
        ADDRESS_BAR: 50,
        HOME_BAR: { x: 136, y: 911, height: 9, width: 156 },
      },
    },
  },
  IPAD: {
    1024: {
      LANDSCAPE: {
        STATUS_BAR: 20,
        ADDRESS_BAR: 50,
        HOME_BAR: { x: 0, y: 0, height: 0, width: 0 },
      },
      PORTRAIT: {
        STATUS_BAR: 20,
        ADDRESS_BAR: 50,
        HOME_BAR: { x: 0, y: 0, height: 0, width: 0 },
      },
    },
    1080: {
      LANDSCAPE: {
        STATUS_BAR: 20,
        ADDRESS_BAR: 50,
        HOME_BAR: { x: 0, y: 0, height: 0, width: 0 },
      },
      PORTRAIT: {
        STATUS_BAR: 20,
        ADDRESS_BAR: 50,
        HOME_BAR: { x: 0, y: 0, height: 0, width: 0 },
      },
    },
    1112: {
      LANDSCAPE: {
        STATUS_BAR: 20,
        ADDRESS_BAR: 50,
        HOME_BAR: { x: 0, y: 0, height: 0, width: 0 },
      },
      PORTRAIT: {
        STATUS_BAR: 20,
        ADDRESS_BAR: 50,
        HOME_BAR: { x: 0, y: 0, height: 0, width: 0 },
      },
    },
    // With Notch
    1133: {
      LANDSCAPE: {
        STATUS_BAR: 24,
        ADDRESS_BAR: 50,
        HOME_BAR: { x: 408, y: 739, height: 9, width: 318 },
      },
      PORTRAIT: {
        STATUS_BAR: 24,
        ADDRESS_BAR: 50,
        HOME_BAR: { x: 234, y: 1118, height: 9, width: 276 },
      },
    },
    1180: {
      LANDSCAPE: {
        STATUS_BAR: 24,
        ADDRESS_BAR: 50,
        HOME_BAR: { x: 431, y: 805, height: 9, width: 318 },
      },
      PORTRAIT: {
        STATUS_BAR: 24,
        ADDRESS_BAR: 50,
        HOME_BAR: { x: 272, y: 1165, height: 9, width: 276 },
      },
    },
    1194: {
      LANDSCAPE: {
        STATUS_BAR: 24,
        ADDRESS_BAR: 50,
        HOME_BAR: { x: 438, y: 819, height: 9, width: 318 },
      },
      PORTRAIT: {
        STATUS_BAR: 24,
        ADDRESS_BAR: 50,
        HOME_BAR: { x: 279, y: 1179, height: 9, width: 276 },
      },
    },
    1366: {
      LANDSCAPE: {
        STATUS_BAR: 24,
        ADDRESS_BAR: 50,
        HOME_BAR: { x: 508, y: 1009, height: 9, width: 350 },
      },
      PORTRAIT: {
        STATUS_BAR: 24,
        ADDRESS_BAR: 50,
        HOME_BAR: { x: 352, y: 1351, height: 9, width: 320 },
      },
    },
  },
};
