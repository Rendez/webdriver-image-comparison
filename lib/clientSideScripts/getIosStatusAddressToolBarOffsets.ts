import { StatusAddressToolBarOffsets } from './statusAddressToolBarOffsets.interfaces';
import { IosOffsets } from '../helpers/constants.interfaces';

/**
 * Get the current height of the iOS status and address bar
 */
export default function getIosStatusAddressToolBarOffsets(iosOffsets: IosOffsets): StatusAddressToolBarOffsets {
  // 1. Determine screen width/height to determine the current iPhone/iPad offset data
  //    For data on the screen sizes check the constants.ts-file
  const { width, height } = window.screen;
  const isIphone = width < 1024 && height < 1024;
  const deviceType = isIphone ? 'IPHONE' : 'IPAD';
  // Need to use matchMedia because the height/size is not always accurate when rotated
  const isLandscape = window.matchMedia('(orientation: landscape)').matches;
  const orientationType = isLandscape ? 'LANDSCAPE' : 'PORTRAIT';
  const defaultPortraitHeight = isIphone ? 667 : 1024;
  const portraitHeight = width > height ? width : height;
  // Not sure if it's a bug, but in Landscape mode the height is the width
  const deviceWidth = isLandscape ? height : width;
  const offsetPortraitHeight =
    Object.keys(iosOffsets[deviceType]).indexOf(portraitHeight.toString()) > -1 ? portraitHeight : defaultPortraitHeight;
  const currentOffsets = iosOffsets[deviceType][offsetPortraitHeight][orientationType];

  // 2. Get the statusbar height
  const statusBarHeight = currentOffsets.STATUS_BAR;
  // 4. Determine the address bar height
  //    Since iOS 15 the address bar for iPhones is at the bottom by default
  //    This is also what we assume because we can't determine it from the
  //    web context
  const osVersion = parseInt(navigator.appVersion.match(/(?:OS |Version\/)(\d+)(?:_|\.)(\d+)(?:_|\.)?(\d+)?/)[1], 10);
  const addressBarOnTop = (!isLandscape && isIphone && osVersion < 15) || isLandscape || !isIphone;
  const statusAddressBarHeight = statusBarHeight + (addressBarOnTop ? currentOffsets.ADDRESS_BAR : 0);
  // 4. Determine the toolbar offsets
  //    In Landscape mode the toolbar is hidden by default and we need to add
  //    home bar data
  const { innerHeight } = window;
  const toolBarHeight = height - innerHeight - statusAddressBarHeight;
  const toolBarOffsets = isLandscape
    ? currentOffsets.HOME_BAR
    : {
        height: toolBarHeight,
        width: deviceWidth,
        x: 0,
        y: height - toolBarHeight,
      };

  // 5. Return the offsets
  return {
    statusAddressBar: {
      height: statusAddressBarHeight,
      width: deviceWidth,
      x: 0,
      y: 0,
    },
    toolBar: toolBarOffsets,
  };
}
