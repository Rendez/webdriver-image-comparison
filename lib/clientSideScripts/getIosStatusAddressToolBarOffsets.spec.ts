import { IOS_DEVICES, NAVIGATOR_APP_VERSIONS, CONFIGURABLE } from '../mocks/mocks';
import getIosStatusAddressToolBarOffsets from './getIosStatusAddressToolBarOffsets';
import { IOS_OFFSETS } from '../helpers/constants';

describe('getIosStatusAddressToolBarOffsets', () => {
  it('should get the correct status, address and toolbar height for a default iPhone with iOS 14 in Portrait mode', () => {
    setEnvironment('IOS', 14, 'IPHONE', false);

    expect(getIosStatusAddressToolBarOffsets(IOS_OFFSETS)).toMatchSnapshot();
  });

  it('should get the correct status, address and toolbar height for a default iPhone with iOS 14 in landscape mode', () => {
    setEnvironment('IOS', 14, 'IPHONE', true);

    expect(getIosStatusAddressToolBarOffsets(IOS_OFFSETS)).toMatchSnapshot();
  });

  it('should get the correct status, address and toolbar height for a default iPhone with iOS 15 in portrait mode', () => {
    setEnvironment('IOS', 15, 'IPHONE_X', false);

    expect(getIosStatusAddressToolBarOffsets(IOS_OFFSETS)).toMatchSnapshot();
  });

  it('should get the correct status, address and toolbar height for a default iPhone with iOS 15 in landscape mode', () => {
    setEnvironment('IOS', 15, 'IPHONE_X', true);

    expect(getIosStatusAddressToolBarOffsets(IOS_OFFSETS)).toMatchSnapshot();
  });

  it('should get the correct status, address and toolbar height for an iPhone with extreme not known dimensions', () => {
    setEnvironment('IOS', 15, 'IPHONE_HEIGHT', false);

    expect(getIosStatusAddressToolBarOffsets(IOS_OFFSETS)).toMatchSnapshot();
  });

  it('should get the correct status, address and toolbar height for an iPad', () => {
    setEnvironment('IOS', 15, 'IPAD', false);

    expect(getIosStatusAddressToolBarOffsets(IOS_OFFSETS)).toMatchSnapshot();
  });

  // it('should get the correct status, address and toolbar height for an iPad with iPadOS 13', () => {
  //   setEnvironment('IPADOS', 13, 'IPAD', false);

  //   expect(getIosStatusAddressToolBarOffsets(IOS_OFFSETS)).toMatchSnapshot();
  // });

  // it('should get the correct status, address and toolbar height for an iPad Pro with to touch id and with iOS 12', () => {
  //   setEnvironment('IOS', 12, 'IPAD_NO_TOUCH', false);

  //   expect(getIosStatusAddressToolBarOffsets(IOS_OFFSETS)).toMatchSnapshot();
  // });
});

/**
 * Set the environment for the test
 */
function setEnvironment(os: string, version: number, phone: string, isLandscape: boolean) {
  // @ts-ignore
  Object.defineProperty(navigator, 'appVersion', { value: NAVIGATOR_APP_VERSIONS[os][version], ...CONFIGURABLE });
  // @ts-ignore
  Object.defineProperty(window.screen, 'width', { value: IOS_DEVICES[phone].width, ...CONFIGURABLE });
  // @ts-ignore
  Object.defineProperty(window.screen, 'height', { value: IOS_DEVICES[phone].height, ...CONFIGURABLE });
  // @ts-ignore
  Object.defineProperty(window, 'innerWidth', { value: IOS_DEVICES[phone].innerWidth, ...CONFIGURABLE });
  // @ts-ignore
  Object.defineProperty(window, 'innerHeight', { value: IOS_DEVICES[phone].innerHeight, ...CONFIGURABLE });
  // @ts-ignore
  Object.defineProperty(window, 'matchMedia', {
    value: jest.fn().mockImplementation(() => ({
      matches: isLandscape,
    })),
    ...CONFIGURABLE,
  });
}
