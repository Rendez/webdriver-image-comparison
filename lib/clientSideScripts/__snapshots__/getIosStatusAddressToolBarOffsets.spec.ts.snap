// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getIosStatusAddressToolBarOffsets should get the correct status, address and toolbar height for a default iPhone with iOS 14 in Portrait mode 1`] = `
Object {
  "statusAddressBar": Object {
    "height": 70,
    "width": 375,
    "x": 0,
    "y": 0,
  },
  "toolBar": Object {
    "height": 44,
    "width": 375,
    "x": 0,
    "y": 623,
  },
}
`;

exports[`getIosStatusAddressToolBarOffsets should get the correct status, address and toolbar height for a default iPhone with iOS 14 in landscape mode 1`] = `
Object {
  "statusAddressBar": Object {
    "height": 50,
    "width": 667,
    "x": 0,
    "y": 0,
  },
  "toolBar": Object {
    "height": 0,
    "width": 0,
    "x": 0,
    "y": 0,
  },
}
`;

exports[`getIosStatusAddressToolBarOffsets should get the correct status, address and toolbar height for a default iPhone with iOS 15 in landscape mode 1`] = `
Object {
  "statusAddressBar": Object {
    "height": 50,
    "width": 812,
    "x": 0,
    "y": 0,
  },
  "toolBar": Object {
    "height": 9,
    "width": 212,
    "x": 300,
    "y": 359,
  },
}
`;

exports[`getIosStatusAddressToolBarOffsets should get the correct status, address and toolbar height for a default iPhone with iOS 15 in portrait mode 1`] = `
Object {
  "statusAddressBar": Object {
    "height": 44,
    "width": 375,
    "x": 0,
    "y": 0,
  },
  "toolBar": Object {
    "height": 133,
    "width": 375,
    "x": 0,
    "y": 679,
  },
}
`;

exports[`getIosStatusAddressToolBarOffsets should get the correct status, address and toolbar height for an iPad 1`] = `
Object {
  "statusAddressBar": Object {
    "height": 74,
    "width": 1024,
    "x": 0,
    "y": 0,
  },
  "toolBar": Object {
    "height": 9,
    "width": 320,
    "x": 352,
    "y": 1351,
  },
}
`;

exports[`getIosStatusAddressToolBarOffsets should get the correct status, address and toolbar height for an iPad with big sizes 1`] = `
Object {
  "statusAddressBar": Object {
    "height": 70,
    "width": 9876,
    "x": 0,
    "y": 0,
  },
  "toolBar": Object {
    "height": 38,
    "width": 9876,
    "x": 0,
    "y": 5394,
  },
}
`;

exports[`getIosStatusAddressToolBarOffsets should get the correct status, address and toolbar height for an iPhone with extreme not known dimensions 1`] = `
Object {
  "statusAddressBar": Object {
    "height": 70,
    "width": 1024,
    "x": 0,
    "y": 0,
  },
  "toolBar": Object {
    "height": 107,
    "width": 1024,
    "x": 0,
    "y": 789,
  },
}
`;
